src/Sylius/Bundle/ShopBundle/Provider/PayumPayResponseProvider.php

```diff
@@ -0,0 +1,127 @@
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Webmozart\Assert\Assert;
+
+final class PayumPayResponseProvider implements PayResponseProviderInterface
+{
+    public function __construct(
+        private Payum $payum,
+    ) {
+    }
+
+    public function getResponse(
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): Response {
+        $payment = $this->getPaymentFromOrder($order);
+        Assert::notNull($payment, 'An existing payment with state "new" must exist.');
+
+        $redirectOptions = $requestConfiguration->getParameters()->get('redirect');
+        if (is_string($redirectOptions)) {
+            $redirectOptions = [
+                'route' => $redirectOptions,
+            ];
+        }
+        $token = $this->provideTokenBasedOnPayment($payment, $redirectOptions);
+
+        return new RedirectResponse($token->getTargetUrl());
+    }
+    public function supports(
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): bool {
+
+        $payment = $this->getPaymentFromOrder($order);
+        if (null === $payment) {
+            return false;
+        }
+
+        $gatewayConfig = $this->getGatewayConfigFromPayment($payment);
+        if (null === $gatewayConfig) {
+            return false;
+        }
+
+        $gatewayName = $gatewayConfig->getGatewayName() ?? '';
+
+        try {
+            $this->payum->getGateway($gatewayName);
+        } catch (InvalidArgumentException) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @param array{route: ?string, parameters: ?string[]} $redirectOptions
+     */
+    private function provideTokenBasedOnPayment(PaymentInterface $payment, array $redirectOptions): TokenInterface
+    {
+        $gatewayConfig = $this->getGatewayConfigFromPayment($payment);
+        Assert::notNull($gatewayConfig, 'An existing gateway config must exist.');
+
+        $config = $gatewayConfig->getConfig();
+        $tokenFactory = $this->payum->getTokenFactory();
+
+        if (isset($config['use_authorize']) && true === (bool) $config['use_authorize']) {
+            $token = $tokenFactory->createAuthorizeToken(
+                $gatewayConfig->getGatewayName(),
+                $payment,
+                $redirectOptions['route'] ?? null,
+                $redirectOptions['parameters'] ?? [],
+            );
+        } else {
+            $token = $tokenFactory->createCaptureToken(
+                $gatewayConfig->getGatewayName(),
+                $payment,
+                $redirectOptions['route'] ?? null,
+                $redirectOptions['parameters'] ?? [],
+            );
+        }
+
+        return $token;
+    }
+
+    private function getPaymentFromOrder(OrderInterface $order): ?PaymentInterface
+    {
+        return $order->getLastPayment(BasePaymentInterface::STATE_NEW);
```

Comment:
Is there need to use const from base payment interface? Values are inherited to Core PaymentInterface and it can be used from it as well.

Reply:
I'm following an advise from the IDE, I wonder why it warns about it and I'm always using the base interface instead of the inherited one.